### ISSUE TITLE
Warning when initializing dataclasses with arguments

### ISSUE THREAD

**Issue by Thlr** [microsoft/pyright#1433]
**Describe the bug**
Pyright prints this warning `[Pyright reportGeneralTypeIssues] [E] Expected no arguments to "MyDataClass" constructor` when passing down arguments that correspond a dataclass attributes.

**To Reproduce**
Create a dataclass
```
@dataclass
class MyDataclass:
    foo: str = field(default="foo")
```

Try to instantiate it

```
MyDataclass(foo="bar")
```

**Expected behavior**
No warnings should be prompted

**VS Code extension or command-line**
OS:
```
❯ uname -r
5.10.10-arch1-1
```
I am running pyright with neovim

`:CocInfo` output:
```
vim version: NVIM v0.4.4
node version: v15.7.0
coc.nvim version: 0.0.80-5cd38e9462
```
I tried with VSCode as well:

![image](https://user-images.githubusercontent.com/37906378/106129689-c54f5f00-6160-11eb-9f1f-d00893debdaa.png)

I installed the `python` and `pyright` extensions

```
❯ code --version
1.52.1
ea3859d4ba2f3e577a159bc91e3074c5d85c0523
x64
```

**Comment by erictraut** [comment]
I'm not able to reproduce the problem you're seeing. Your sample code doesn't generate an error for me.

```python
from dataclasses import dataclass, field

@dataclass
class MyDataclass:
    foo: str = field(default="foo")

MyDataclass(foo="bar")
```

**Comment by Thlr** [comment]
Indeed. I wasn't able to reproduce it as well, I thought giving an equivalent snippet of code would do the trick. Yet, it does happen on the code I am working on as you can see on the screenshot.

We use poetry, do you think it could come from there?

Here are the dependencies in the `pyproject.toml`

```
[tool.poetry.dependencies]
python = "^3.7"
aiohttp = "^3.6.2"
pyyaml = "^5.3.1"
kubernetes-asyncio = "^12"
marshmallow = "^3.8.0"
marshmallow-dataclass = "^8.0.0"
fqdn = "^1.4.0"
importlib-metadata = {version = "^1.0", python = "<3.8"}

typer = "^0.3.2"
pygments = "^2.7.1"
coloredlogs = "^14.0"
packaging = "^20.8"
[tool.poetry.dev-dependencies]
# ipython = "*"
pytest = "*"
black = "^20.8b1"
isort = "^5.5.4"
flake8 = "^3.8.3"
pylint = "^2.6.0"
recommonmark = "^0.6.0"
mypy = "^0.782"
```

**Comment by erictraut** [comment]
I don't know much about the poetry package, but from what I've read about it, it seems unlikely that it has anything to do with this issue.

If you're unable to provide a repro, I'm going to mark this as closed. If you're able to find a repro, please add it to this issue, and I'll reopen it.

**Comment by Thlr** [comment]
Of course. Thank you for your time.

**Comment by Delwin9999** [comment]
I'm getting the same problem and the difference is that I only get it when set to python 3.6 and use the dataclasses backport package.

**Comment by Delwin9999** [comment]
![pyright_dataclass_working](https://user-images.githubusercontent.com/817061/122487906-2b624480-cf91-11eb-9ad9-acabf5c342f1.jpg)
![pyright_dataclass_failing](https://user-images.githubusercontent.com/817061/122487910-2dc49e80-cf91-11eb-96cf-4ae161f4827b.jpg)
